src/typeorm-aurora-data-api-driver.ts

```diff
@@ -1,95 +1,113 @@
-
-    let newQueryString = ''
-    let currentQuote = null
-
-    for (let i = 0; i < query.length; i += 1) {
-      const currentCharacter = query[i]
-      const currentCharacterEscaped = i !== 0 && query[i - 1] === '\\'
-
-      if (currentCharacter === '?' && !currentQuote) {
+const transformQuery = (query: string, parameters?: any[]): [string, number] => {
+  const quoteCharacters = ['\'', '"']
+
+  let newQueryString = ''
+  let currentQuote = null
+  let numberOfParametersInQueryString = 0
+  let extendedParameterCount = 0
+
+  for (let i = 0; i < query.length; i += 1) {
+    const currentCharacter = query[i]
+    const currentCharacterEscaped = i !== 0 && query[i - 1] === '\\'
+
+    if (currentCharacter === '?' && !currentQuote) {
+      const parameter = parameters![numberOfParametersInQueryString]
+      if (Array.isArray(parameter)) {
+        parameter.forEach((_, index) => {
+          const isLastParameter = index === parameter.length - 1
+          const paramName = `:e_param_${extendedParameterCount}${isLastParameter ? '' : ', '}`
+          extendedParameterCount += 1
+          newQueryString += paramName
+        })
+      } else {
         const paramName = `:param_${numberOfParametersInQueryString}`
-
         numberOfParametersInQueryString += 1
         newQueryString += paramName
-      } else if (quoteCharacters.includes(currentCharacter) && !currentCharacterEscaped) {
-        if (!currentQuote) {
-          currentQuote = currentCharacter
-        } else if (currentQuote === currentCharacter) {
-          currentQuote = null
-        }
-
-        newQueryString += currentCharacter
-      } else {
-        newQueryString += currentCharacter
       }
+
+      continue
     }
 
-    if (
-      parameters &&
-      parameters.length > 0 &&
-      parameters.length % numberOfParametersInQueryString !== 0
-    ) {
-      throw new Error(`Number of parameters mismatch, got ${numberOfParametersInQueryString} in query string \
-            and ${parameters.length} in input`)
+    if (quoteCharacters.includes(currentCharacter) && !currentCharacterEscaped) {
+      if (!currentQuote) {
+        currentQuote = currentCharacter
+      } else if (currentQuote === currentCharacter) {
+        currentQuote = null
+      }
     }
 
-    const transformedParameters: any[] = []
+    newQueryString += currentCharacter
+  }
+
+  return [newQueryString, numberOfParametersInQueryString]
+}
+
+const transformParameters = (numberOfParametersInQueryString: number, parameters?: any[]) => {
+  if (
+    parameters &&
+    parameters.length > 0 &&
+    parameters.length % numberOfParametersInQueryString !== 0
+  ) {
+    throw new Error(
+      `Number of parameters mismatch, got ${numberOfParametersInQueryString} in query string \
+      and ${parameters.length} in input`)
+  }
 
-    if (parameters && parameters.length > 0) {
-      const numberOfObjects = parameters.length / numberOfParametersInQueryString
+  let extendedParameterCount = 0
```

Comment:
Not sure we need that here since we have an index already and we know the count

Reply:
The reason I do this is that I count the extended parameters independently of the number of parameters in the query string.

As far as I am concerned the number of parameters is the same, but one of them is extended. I chose to use a different parameter name for extended array parameters (```e_param_X```) so that it's very obvious in the logs that these parameters have been extended.

I still return the same number of parameters in the query string as you did previously. 

The index is therefore present so it has it's own seed and takes into account there could of course be more than one array being expanded in any given query.

Reply:
I think we shouldn't distinguish parameters and extended parameters and return the number of extended parameters as a number of parameters from `transformQuery`. That way we don't need to count them again which will simplify `transformParameters`. It shouldn't break the code as I think and we have tests to check for that.